src/back/hlsl/help.rs

```diff
@@ -413,6 +443,335 @@ impl<'a, W: Write> super::Writer<'a, W> {
         Ok(())
     }
 
+    pub(super) fn write_wrapped_struct_matrix_get_function_name(
+        &mut self,
+        access: WrappedStructMatrixAccess,
+    ) -> BackendResult {
+        let name = &self.names[&NameKey::Type(access.ty)];
+        let field_name = &self.names[&NameKey::StructMember(access.ty, access.index)];
+        write!(self.out, "GetMat{}On{}", field_name, name)?;
+        Ok(())
+    }
+
+    /// Writes a function used to get a matCx2 from within a structure.
+    pub(super) fn write_wrapped_struct_matrix_get_function(
+        &mut self,
+        module: &crate::Module,
+        access: WrappedStructMatrixAccess,
+    ) -> BackendResult {
+        use crate::back::INDENT;
+
+        const STRUCT_ARGUMENT_VARIABLE_NAME: &str = "obj";
+
+        // Write function return type and name
+        let member = match module.types[access.ty].inner {
+            crate::TypeInner::Struct { ref members, .. } => &members[access.index as usize],
+            _ => unreachable!(),
+        };
+        let ret_ty = &module.types[member.ty].inner;
+        self.write_value_type(module, ret_ty)?;
+        write!(self.out, " ")?;
+        self.write_wrapped_struct_matrix_get_function_name(access)?;
+
+        // Write function parameters
+        write!(self.out, "(")?;
+        let struct_name = &self.names[&NameKey::Type(access.ty)];
+        write!(
+            self.out,
+            "{} {}",
+            struct_name, STRUCT_ARGUMENT_VARIABLE_NAME
+        )?;
+
+        // Write function body
+        writeln!(self.out, ") {{")?;
+
+        // Write return value
+        write!(self.out, "{}return ", INDENT)?;
+        self.write_value_type(module, ret_ty)?;
+        write!(self.out, "(")?;
+        let field_name = &self.names[&NameKey::StructMember(access.ty, access.index)];
+        match module.types[member.ty].inner {
+            crate::TypeInner::Matrix { columns, .. } => {
+                for i in 0..columns as u8 {
+                    write!(
+                        self.out,
+                        "{}.{}_{}",
+                        STRUCT_ARGUMENT_VARIABLE_NAME, field_name, i
+                    )?;
+                    if i < columns as u8 - 1 {
```

Comment:
small trick: if you move this condition to the beginning of the loop body (instead of the end), you can just compare `i != 0`